crates/sui-core/src/authority.rs

```diff
@@ -360,39 +366,61 @@ impl AuthorityState {
         confirmation_transaction: ConfirmationTransaction,
     ) -> SuiResult<TransactionInfoResponse> {
         self.metrics.total_certs.inc();
-        let transaction_digest = *confirmation_transaction.certificate.digest();
-
-        // Ensure an idempotent answer.
-        if self.database.effects_exists(&transaction_digest)? {
-            let info = self.make_transaction_info(&transaction_digest).await?;
-            debug!("Transaction {transaction_digest:?} already executed");
-            return Ok(info);
-        }
+        let certificate = confirmation_transaction.certificate;
+        let transaction_digest = *certificate.digest();
 
-        if self.halted.load(Ordering::SeqCst)
-            && !confirmation_transaction
-                .certificate
-                .data
-                .kind
-                .is_system_tx()
-        {
+        if self.halted.load(Ordering::SeqCst) && !certificate.data.kind.is_system_tx() {
             // TODO: Do we want to include the new validator set?
             return Err(SuiError::ValidatorHaltedAtEpochEnd);
         }
 
         // Check the certificate and retrieve the transfer data.
         tracing::trace_span!("cert_check_signature")
-            .in_scope(|| {
-                confirmation_transaction
-                    .certificate
-                    .verify(&self.committee.load())
-            })
+            .in_scope(|| certificate.verify(&self.committee.load()))
             .map_err(|e| {
                 self.metrics.signature_errors.inc();
                 e
             })?;
 
-        self.process_certificate(confirmation_transaction).await
+        // This acquires a lock on the tx digest to prevent multiple concurrent executions of the
+        // same tx. While we don't need this for safety (tx sequencing is ultimately atomic), it is
+        // very common to receive the same tx multiple times simultaneously due to gossip, so we
+        // may as well hold the lock and save the cpu time for other requests.
+        //
+        // Note that this lock has some false contention (since it uses a MutexTable), so you can't
+        // assume that different txes can execute concurrently. This is probably the fastest way
```

Comment:
False contention == Two different TXes digests that hash to the same slot in the MutexTable. It would be correct for them to execute concurrently, but the design prevents it. I think it's a worthwhile trade off because with this design we can easily drive contention as low as we want, and as the saying goes, locks aren't slow, lock contention is slow. I will add a quick definition of false contention here.

Re futures: you can't easily cache futures in rust because of lifetime hell. The way to do it is to create a tokio::sync::broadcast channel that everybody waits on. This is way more complicated and probably slower than just holding a lock.

Reply:
:(